### ISSUE TITLE
Fix (benchmarking-related?) CI issues

### ISSUE THREAD

**Issue by brainstorm** [umccr/htsget-rs#90]
https://github.com/umccr/htsget-rs/runs/6725133792?check_suite_focus=true#step:11:294

It seems that criterion is trying to fish out some prior result to compare against?: https://github.com/umccr/htsget-rs/runs/6725133792?check_suite_focus=true#step:13:20

Could also be related with my recent await-ification of many of the tests. I'll investigate.

Future proofing warning worth fixing too: https://github.com/umccr/htsget-rs/runs/6725133792?check_suite_focus=true#step:11:139

Also need to investigate on whether to run the benchmarks with `--release` mode on as it's the production baseline we want to compare against.

Also when running the benchmarks locally, they don't seem to finish?:

`cargo bench --bench search-benchmarks --bench request-benchmarks -- LIGHT --output-format bencher | tee output.txt`

**Comment by brainstorm** [comment]
Real problem is that it doesn't seem to connect to the already spawned server (Connection refused)? It works fine locally though: 

```
(...)
Status: Downloaded newer image for ga4gh/htsget-refserver:1.5.0
docker.io/ga4gh/htsget-refserver:1.5.0
Server started on port 3000!
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: reqwest::Error { kind: Request, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(8080), path: "/reads/bam/htsnexus_test_NA12878", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })) }', htsget-http-actix/benches/request_benchmarks.rs:52:61
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: bench failed
```

**Comment by brainstorm** [comment]
Local run is fine:

```
% time cargo bench --bench search-benchmarks --bench request-benchmarks -- LIGHT --output-format bencher
    Finished bench [optimized] target(s) in 0.44s
     Running benches/request_benchmarks.rs (target/release/deps/request_benchmarks-61adb651982bafd5)
Gnuplot not found, using plotters backend
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `target/debug/htsget-http-actix`
1.5.0: Pulling from ga4gh/htsget-refserver
Digest: sha256:b93ab0593f58165351a136f19661228aec203ccaa74d746bd309e8936d8038de
Status: Image is up to date for ga4gh/htsget-refserver:1.5.0
docker.io/ga4gh/htsget-refserver:1.5.0
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Server started on port 3000!
test Requests/[LIGHT] simple request htsget-rs ... bench:    50192597 ns/iter (+/- 2604779)
test Requests/[LIGHT] simple request htsget-refserver ... bench:   241670291 ns/iter (+/- 52979001)
test Requests/[LIGHT] with region htsget-rs ... bench:    51889638 ns/iter (+/- 1895549)

Warning: Unable to complete 100 samples in 30.0s. You may wish to increase target time to 46.1s, or reduce sample count to 60.
test Requests/[LIGHT] with region htsget-refserver ... bench:   431346333 ns/iter (+/- 9810885)
test Requests/[LIGHT] with two regions htsget-rs ... bench:    95558263 ns/iter (+/- 4413729)

Warning: Unable to complete 100 samples in 30.0s. You may wish to increase target time to 58.0s, or reduce sample count to 50.
test Requests/[LIGHT] with two regions htsget-refserver ... bench:   573878041 ns/iter (+/- 11200399)

Warning: Unable to complete 100 samples in 30.0s. You may wish to increase target time to 52.9s, enable flat sampling, or reduce sample count to 50.
test Requests/[LIGHT] with VCF htsget-rs ... bench:    10081166 ns/iter (+/- 327103)
test Requests/[LIGHT] with VCF htsget-refserver ... bench:   184615812 ns/iter (+/- 3240979)

     Running benches/search_benchmarks.rs (target/release/deps/search_benchmarks-2731ff10a82d3df1)
Gnuplot not found, using plotters backend
test Queries/[LIGHT] Bam query all ... bench:     4387442 ns/iter (+/- 193169)
test Queries/[LIGHT] Bam query specific ... bench:     5329601 ns/iter (+/- 321501)
test Queries/[LIGHT] Bam query header ... bench:      356823 ns/iter (+/- 71399)
test Queries/[LIGHT] Cram query all ... bench:      402423 ns/iter (+/- 59736)
test Queries/[LIGHT] Cram query specific ... bench:      658649 ns/iter (+/- 454257)
test Queries/[LIGHT] Cram query header ... bench:      329839 ns/iter (+/- 145571)
test Queries/[LIGHT] Vcf query all ... bench:      399011 ns/iter (+/- 15657)
test Queries/[LIGHT] Vcf query specific ... bench:      520640 ns/iter (+/- 10706)
test Queries/[LIGHT] Vcf query header ... bench:      226984 ns/iter (+/- 15869)
test Queries/[LIGHT] Bcf query all ... bench:      427458 ns/iter (+/- 20275)
test Queries/[LIGHT] Bcf query specific ... bench:      572761 ns/iter (+/- 30455)
test Queries/[LIGHT] Bcf query header ... bench:      250903 ns/iter (+/- 14151)

782.85s user 136.34s system 25% cpu 59:13.80 total
```

**Comment by mmalenic** [comment]
This almost feels like a concurrency, race-condition type of error, however it's not obvious where it is occuring. It seems to be related to generating self-signed certificates, e.g. here: https://github.com/umccr/htsget-rs/runs/6723277672?check_suite_focus=true#step:11:179.

The certificates are generated here, and placed into a temporary directory: https://github.com/umccr/htsget-rs/blob/75aed4760755a13d83e819fbcacbc77f40ee68b2/htsget-http-actix/benches/request_benchmarks.rs#L140-L141

But this should all be sequential processing, so I think it's something to do with how TempDir interacts with writing files. I know there's a bit of weirdness with dropping a reference to the directory clearing its contents. I think the base_path should be passed into the function so that it lives long enough, because maybe the spawned process doesn't have enough time to grab hold of the certificate before it is cleared.

As a side note, I think the github action should fail if either the spawned htsget-rs or htsget-refserver process fails with an error, as this would allow for easier debugging.

**Comment by brainstorm** [comment]
PR #94 seems to fix this issue, but issue #93 seems to persist, even if it really points to the same underlying race condition... most probably the CLI has a different codepaths and/or there's some duplicated code?
